### PR TITLE
DAML-LF: utility to compute duplicated keys in a transaction.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
The Engine does not verify key duplication when construction a
transaction, this task is currently let to the ledger implementation.
This PR provides an utility function to comptute the duplicated keys
in a transaction.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by simonmaxen-da** [comment]
Looks almost identical to what is [currently being used for daml-on-corda](https://github.com/DACH-NY/daml-on-corda/blob/16d4270ca8cd055abe752350cb6845fbfa2e0960/engine/src/main/scala/com/digitalasset/platform/corda/validation/DamlStateValidator.scala#L184) so I'm happy.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 1]
```diff
@@ -466,6 +466,42 @@ object GenTransaction extends value.CidContainer3[GenTransaction] {
...
+  @throws[IllegalArgumentException]
+  def duplicatedContractKeys(tx: VersionedTransaction[NodeId, Value.ContractId]): Set[GlobalKey] = {
+
+    import GlobalKey.{assertBuild => globalKey}
+
+    case class State(active: Set[GlobalKey], duplicates: Set[GlobalKey]) {
+      def created(key: GlobalKey): State =
+        if (active(key)) copy(duplicates = duplicates + key) else copy(active = active + key)
+      def consumed(key: GlobalKey): State =
+        if (active(key)) copy(active = active - key) else this
```
Comment (simonmaxen-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:480]: Could drop active check for this an referenced as set operation benign in excluded case.

**Review by simonmaxen-da — APPROVED** [review]

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:480]: This check just avoids unnecessary allocation of `State` object.

**Review by oliverse-da — APPROVED** [review]

**Comment by simonmaxen-da** [comment]
@remyhaemmerle-da - nice one, I've added using this to the DAML 1.5 upgrade todo.

**Review by andreaslochbihler-da — APPROVED** [review]
Looks good. We'll use this in Canton when Canton pulls in this PR with the next snapshot release.

**Pull request merged**